### DISCUSSION TITLE
Best practice to store upto 5 million records

### DISCUSSION THREAD

**Discussion by fahadmalikawan** [exadel-inc/CompreFace#1070] [Q&A]
I have to implement the system where lot of Face Detection Cameras are installed. 
These cameras will throw the images into the server's FTP folder.
Compreface System will be installed in the server containing registered persons i.e. Citizens, employees etc.
A utility will call compreface API to verify the FTP pictures. 
The faces not found in the Compreface server will be marked strangers and will be **stored as subjects in the Compreface system**

Now these strangers might be a very large quanitiy.
I need guidance to implement the system smartly
I have powerful computer. And can install any recommended GPU. Currently I have RTX 3060. But i can install 3090 if recommended

Questions in my mind:

1. What is the practical limit of faces in an application or server?
2.  Storing millions of faces in single Compreface Application will affect the performance dramatically?
3. Do i need to make different applications eg one application for 1 million records, and manage in such a way that after all the applications are loaded, i clear the subject from the first application. so that can keep the data for a couple of months.
4. Storing subjects without saving images to the Database will result in faster recognition?
5. **(Important)** I will have different applications holding different data. Eg Residents and Citizens on one application. Blacklisted people in other application. Strangers in other multiple application. So Now I found that currently the main time taking task is extraction of embeddings of a face, not the recognition. 
**The Question**: If we want to compare a single face in different applications, is there a way that we extract face embeddings calculations once and compare those embeddings in all the applications?

Will be grateful for any guidance. Thanks

**Comment by pospielov** [comment]
1. Practical limit is about 100 000 - 200 000 examples. Statistics show at least two cases of using it with more than 1 000 000 examples, but it's not practical from an effective RAM usage perspective.
2. Comparing embeddings is optimized for tens of thousands of examples. The biggest issue with recognizing millions of faces is RAM. I believe performance degradation for 1 million records won't be critical, but it's definitely not optimal.
3. Probably, it will help if you split them by 200 000 examples. Of course, if you won't have more than 200 000 examples per month.
4. Stored images are used only to show them on UI, so no, they won't affect performance. But I still recommend disabling storing them as it will save you a lot of storage space.
5. We added it in the 1.2 version but didn't release it yet - https://github.com/exadel-inc/CompreFace/blob/master/docs/Rest-API-description.md#recognition-and-verification-using-embedding. Not sure about estimates of 1.2 release - we are trying to add support of Nvidia 4xxx GPUs, but we faced lots of issues when we try to update libraries. So we will release 1.2 release as soon as we find the solution. 

When we created the current CompreFace architecture, we optimized it for small face collections, this is why your case is not very suitable for CompreFace. 
In parallel, we are developing a private version - CompreFace Enterprise -  with the support of a vector database, that allows us to increase practical limits drastically - I tested it with 20 million records on my laptop. If you are interested in it, please write us an email on [compreface.support­@exadel.com](mailto:compreface.support@exadel.com)
